fontforge/plugin.c

```diff
@@ -0,0 +1,558 @@
+                if (modname == NULL) {
+                    LogError(_("No module name for '%s' in plugin config -- skipping.\n"), groups[i]);
+                    continue;
+                }
+                char *pkgname = g_key_file_get_string(conf, groups[i], "Package name", NULL);
+                char *sm_string = g_key_file_get_string(conf, groups[i], "Active", NULL);
+                char *url = g_key_file_get_string(conf, groups[i], "URL", NULL);
+                PluginEntry *pe = NewPluginEntry(copy(groups[i]), pkgname, modname, url, pluginStartupModeFromString(sm_string));
+                free(sm_string);
+                plugin_data = g_list_append(plugin_data, pe);
+            }
+            g_strfreev(groups);
+        }
+        free(fname);
+        free(pdir);
+    }
+    g_key_file_free(conf);
+}
+
+void LoadPlugin(PluginEntry *pe) {
+    PyObject *str, *tmp, *tmp2;
+    if (!use_plugins || !pe->is_present || pe->pyobj != NULL || pe->entrypoint == NULL) {
+        return;
+    }
+    str = PyUnicode_FromString("load");
+    pe->pyobj = PyObject_CallMethodObjArgs(pe->entrypoint, str, NULL);
+    Py_DECREF(str);
+    if (pe->pyobj != NULL) {
+        tmp = PyObject_GetAttrString(pe->pyobj, "fontforge_plugin_init");
+        if (tmp != NULL && PyFunction_Check(tmp)) {
+            PyObject *args = PyTuple_New(0);
+            PyObject *kwargs = PyDict_New();
+            char *b1 = getPluginDirName(), *b2 = smprintf("%s/%s", b1, pe->name);
+            PyObject *dstr = PyUnicode_FromString(b2);
+            free(b1);
+            free(b2);
+            PyDict_SetItemString(kwargs, "preferences_path", dstr);
+            tmp2 = PyObject_Call(tmp, args, kwargs);
+            if (tmp2 == NULL) {
+                LogError(_("Skipping plugin %s: module '%s': Error calling 'fontforge_plugin_init' function\n"), pe->name, pe->module_name);
+                PyErr_Print();
+            } else {
+                pe->is_well_formed = true;
+                Py_DECREF(tmp2);
+            }
+            Py_DECREF(dstr);
+            Py_DECREF(kwargs);
+            Py_DECREF(args);
+            Py_DECREF(tmp);
+        } else {
+            LogError(_("Skipping plugin %s: module '%s': Lacks 'fontforge_plugin_init' function\n"), pe->name, pe->module_name);
+            if (tmp != NULL) {
+                Py_DECREF(tmp);
+            } else {
+                PyErr_Clear();
+            }
+        }
+        tmp = PyObject_GetAttrString(pe->pyobj, "fontforge_plugin_config");
+        pe->has_prefs = (tmp != NULL && PyFunction_Check(tmp));
+        if (tmp != NULL) {
+            Py_DECREF(tmp);
+        } else {
+            PyErr_Clear();
+        }
+    } else {
+        LogError(_("Skipping plugin %s: module '%s': Could not load.\n"), pe->name, pe->module_name);
+        PyErr_Print();
+    }
+    Py_DECREF(pe->entrypoint);
+    pe->entrypoint = NULL;
+}
+
+static void ReimportPlugins() {
```

Comment:
When would you want to reimport? How well does this work in practice? I seem to recall issues with actually reloading already imported code.

Reply:
The dialog provides a means of loading a specific plugin. The python API does not -- all you can do to load a plugin after enabling it with `configurePlugins()` is call `loadPlugins()` -- possibly for a second, third, etc. time. 

The internal plumbing should be such that each plugin is only loaded once. Hence the `pe->is_present && pe->pyobj == NULL` in the guard. However, I do now see a problem with this function in that it doesn't appear to be checking the `startup_mode`. I'll look into that.

Reply:
Added the `startup_mode` check.